### PR TITLE
backport: bitcoin#18790, #19164, #19226, #19276, #19348, #19422, #19840, #20071, #20076, partial #19041

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
Just regular bitcoin backports, mostly from v21

## What was done?
 - partial bitcoin/bitcoin#19041
 - bitcoin/bitcoin#19164
 - bitcoin/bitcoin#19226
 - bitcoin/bitcoin#19276
 - bitcoin/bitcoin#19348
 - bitcoin/bitcoin#19422
 - bitcoin/bitcoin#19840
 - bitcoin/bitcoin#18790
 - bitcoin/bitcoin#20071
 - bitcoin/bitcoin#20076

## How Has This Been Tested?
Run unit/functional tests

## note for reviewers: 
 - There's some difference with original PR bitcoin#20071 but that's not important because we do not use travis anyhow.
Moreover, the variable TRAVIS_BRANCH would be removed anyway in bitcoin#20697 - let's just skip it for simplicity
 - clang in our CI gets linkage error if try to actually do linkage with libc++; maybe that's too old ubuntu, maybe something else - difficult to debug it with remote server. So, bitcoin#19041 is partial.
 - 
## Breaking Changes
N/A

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -2360,20 +2360,15 @@ bool AppInitMain(const CoreContext& context, NodeContext& node, interfaces::Bloc
...
-
-            std::string strCmd = block_notify;
-            if (!strCmd.empty()) {
-                ReplaceAll(strCmd, "%s", pBlockIndex->GetBlockHash().GetHex());
-                std::thread t(runCommand, strCmd);
-                t.detach(); // thread runs free
-            }
-        };
-        uiInterface.NotifyBlockTip_connect(BlockNotifyCallback);
+    const std::string block_notify = args.GetArg("-blocknotify", "");
```
Comment (UdjinM6) [inline src/init.cpp:2363]: 19840: would probably make sense to also fix `-chainlocknotify` while at it (in a follow-up commit maybe)

File: ci/lint/04_install.sh [hunk 2]
```diff
@@ -16,6 +16,6 @@ travis_retry pip3 install vulture==2.3
 travis_retry pip3 install yq
 travis_retry pip3 install mypy==0.700
 
-SHELLCHECK_VERSION=v0.6.0
-curl -s "https://storage.googleapis.com/shellcheck/shellcheck-${SHELLCHECK_VERSION}.linux.x86_64.tar.xz" | tar --xz -xf - --directory /tmp/
+SHELLCHECK_VERSION=v0.7.1
```
Comment (UdjinM6) [inline ci/lint/04_install.sh:19]: 19348: missing `mypy` bump to `0.781` above

File: src/init.cpp [hunk 3]
```diff
@@ -2360,32 +2360,24 @@ bool AppInitMain(const CoreContext& context, NodeContext& node, interfaces::Bloc
...
+        uiInterface.NotifyBlockTip_connect([block_notify](SynchronizationState sync_state, const CBlockIndex* pBlockIndex) {
+            if (sync_state != SynchronizationState::POST_INIT || !pBlockIndex) return;
+            std::string command = block_notify;
+            ReplaceAll(command, "%s", pBlockIndex->GetBlockHash().GetHex());
+            std::thread t(runCommand, command);
+            t.detach(); // thread runs free
+        });
+    }
+    const std::string chainlock_notify = args.GetArg("-chainlocknotify", "");
+    if (chainlock_noify.empty()) {
```
Comment (UdjinM6) [inline src/init.cpp:2374]: 19840:
```suggestion
    if (!chainlock_notify.empty()) {
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**